File: web/app/lib/document-content-review.ts
/** Open analyst notes with source passages rechecked before display. */
import { CORPUS_PREFIX, parseCorpusRevision, readVerifiedPage, type CorpusBucket, type CorpusRevision, type FilingIdentity } from "./document-corpus";

const record = (v: unknown): v is Record<string, unknown> => typeof v === "object" && v !== null && !Array.isArray(v);
const box = (v: unknown): v is number[] => Array.isArray(v) && v.length === 4 && v.every(n => typeof n === "number" && Number.isFinite(n)) && v[0] < v[2] && v[1] < v[3];
export type ContentReview = { id: string; kind: "source_review"; title: string; note: string; status: "open";
  semantic_verification: "not_performed"; points: { page: number; bbox: number[]; source_span_ids: number[];
    source_text: string; source_page_artifact_sha256: string }[] };

export async function getContentReviews(bucket: CorpusBucket, filing: FilingIdentity, revision: CorpusRevision): Promise<ContentReview[]> {
  const object = await bucket.get(`${CORPUS_PREFIX}filings/${filing.bank_ticker}/${filing.period}/${filing.kind}.json`);
  if (!object || object.size > 8_000_000) throw new Error("Missing or oversized review index");
  const index: unknown = await object.json();
  const current = parseCorpusRevision(index, filing);
  if (!current || current.source.pdf_sha256 !== revision.source.pdf_sha256 || current.evidence_key !== revision.evidence_key) {
    throw new Error("Source changed while loading its content reviews");
  }
  const receipt = record(index) && record(index.resume_receipt) ? index.resume_receipt : null;
  const benchmark = receipt && record(receipt.source_benchmark) ? receipt.source_benchmark : null;
  if (!benchmark) return [];
  if (!Array.isArray(benchmark.checks) || benchmark.checks.length > 100) throw new Error("Invalid review checks");
  const findings: unknown[] = [];
  for (const check of benchmark.checks) {
    if (!record(check)) throw new Error("Invalid review check");
    if (check.content_reviews === undefined) continue;
    if (check.passed !== true || !Array.isArray(check.content_reviews)) throw new Error("Unverified review references");
    findings.push(...check.content_reviews);
  }
  if (!findings.length) return [];
  if (!receipt || receipt.schema_version !== "corpus-receipt-1" || benchmark.status !== "passed"
      || benchmark.scope !== "annotated_cases_only" || findings.length > 100
      || revision.evidence_key !== `${CORPUS_PREFIX}sources/${revision.source.pdf_sha256}/${receipt.evidence_artifact_sha256}.jsonl.gz`) {
    throw new Error("Content reviews differ from the current native source");
  }
  const pages = new Map<number, Awaited<ReturnType<typeof readVerifiedPage>>>();
  const ids = new Set<string>();
  for (const finding of findings) {
    if (!record(finding) || finding.kind !== "source_review" || finding.status !== "open"
        || finding.semantic_verification !== "not_performed"
        || !["id", "title", "note"].every(k => typeof finding[k] === "string" && finding[k].trim())
        || !Array.isArray(finding.points) || finding.points.length < 2 || finding.points.length > 20
        || ids.has(String(finding.id))) throw new Error("Invalid open content review");
    ids.add(String(finding.id));
    for (const point of finding.points) {
      if (!record(point) || typeof point.page !== "number" || !Number.isSafeInteger(point.page)
          || point.page < 1 || point.page > revision.page_count || !box(point.bbox)
          || typeof point.source_text !== "string" || !Array.isArray(point.source_span_ids)
          || !point.source_span_ids.length || !point.source_span_ids.every((n, i, all) => typeof n === "number"
            && Number.isSafeInteger(n) && n >= 0 && (i === 0 || n > all[i - 1]))) throw new Error("Invalid review source point");
      if (!pages.has(point.page)) {
        const source = await bucket.get(revision.evidence_key);
        if (!source) throw new Error("Missing native review evidence");
        pages.set(point.page, await readVerifiedPage(source.body, point.page, revision.source.pdf_sha256, revision.page_count, "source"));
      }
      const saved = pages.get(point.page)!;
      if (!Array.isArray(saved.manifest.page_sha256) || saved.manifest.page_sha256[point.page - 1] !== point.source_page_artifact_sha256
          || !Array.isArray(saved.page.spans)) throw new Error("Review page checksum mismatch");
      const spans = point.source_span_ids.map(id => (saved.page.spans as unknown[]).filter(s => record(s) && s.id === id));
      const bounds = point.bbox;
      if (spans.some(s => s.length !== 1) || spans.some(([s]) => !record(s) || !box(s.bbox)
          || s.bbox[0] < bounds[0] || s.bbox[1] < bounds[1] || s.bbox[2] > bounds[2] || s.bbox[3] > bounds[3])
          || spans.map(([s]) => (s as Record<string, unknown>).text).join("\n") !== point.source_text) {
        throw new Error("Review passage differs from retained source spans");
      }
    }
  }
  return findings as ContentReview[];
}

File: src/audit_reports/document_content_review.py
"""Source-bound open review notes; never corrected values or semantic approval."""
from __future__ import annotations

import hashlib
import math

from .document_evidence import _canonical_json


def validate_content_review(case: dict, pages: dict[int, dict]) -> bool:
    """Require exact retained page bytes, span occurrences, wording and geometry.

    The note is a reviewer's interpretation. Only its source references are
    checked here. No comparison of accounting meanings is inferred from this.
    """
    if (case.get("status") != "open" or case.get("semantic_verification") != "not_performed"
            or not all(isinstance(case.get(k), str) and case[k].strip() for k in ("id", "title", "note"))
            or not isinstance(case.get("points"), list) or not 2 <= len(case["points"]) <= 20):
        return False
    for point in case["points"]:
        if not isinstance(point, dict) or type(point.get("page")) is not int or point["page"] not in pages:
            return False
        page = pages[point["page"]]
        if hashlib.sha256(_canonical_json(page).encode("utf-8")).hexdigest() != point.get("source_page_artifact_sha256"):
            return False
        ids, box = point.get("source_span_ids"), point.get("bbox")
        if (not isinstance(ids, list) or not ids or any(type(i) is not int for i in ids)
                or ids != sorted(set(ids)) or not isinstance(box, list) or len(box) != 4
                or any(type(v) not in (int, float) or not math.isfinite(v) for v in box)
                or box[0] >= box[2] or box[1] >= box[3]):
            return False
        spans = {s["id"]: s for s in page["spans"]}
        if any(i not in spans for i in ids):
            return False
        selected = [spans[i] for i in ids]
        if "\n".join(s["text"] for s in selected) != point.get("source_text"):
            return False
        if any(not (box[0] <= s["bbox"][0] <= s["bbox"][2] <= box[2]
                    and box[1] <= s["bbox"][1] <= s["bbox"][3] <= box[3]) for s in selected):
            return False
    return True
